### PR TITLE
Add tests to SingleThreadedSchedulerClient() and document the memory model

### PR DESCRIPTION
**Pull request by skeees**
As discussed in #13023 I've split this test out into a separate pr

This test (and documentation update) makes explicit the guarantee (previously undefined, but implied by the 'SingleThreaded' in `SingleThreadedSchedulerClient()`) - that callbacks pushed to the `SingleThreadedSchedulerClient()` obey the single threaded model for memory and execution - specifically, the callbacks are executed fully and in order, and even in cases where a subsequent callback is executed by a different thread, sequential consistency of memory for all threads executing these callbacks is maintained.

Maintaining memory consistency should make the api more developer friendly - especially for users of the validationinterface. To the extent that there are performance implications from this decision, these are not currently present in practice because all use of this scheduler happens on a single thread currently, furthermore the lock should guarantee consistency across callback executions even when callbacks are executed by multiple threads (as the test does).

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Can you also note the memory model in validationinterface, since that's the "real" API here.

**Comment by skeees** [comment]
Updated

File: src/test/scheduler_tests.cpp [hunk 1]
```diff
@@ -112,4 +112,46 @@ BOOST_AUTO_TEST_CASE(manythreads)
...
+    // each queue should be well ordered with respect to itself but not other queues
+    SingleThreadedSchedulerClient queue1(&scheduler);
+    SingleThreadedSchedulerClient queue2(&scheduler);
+
+    // create more threads than queues
+    // if the queues only permit execution of one task at once then
+    // the extra threads should effectively be doing nothing
+    // if they don't we'll get out of order behaviour
+    boost::thread_group threads;
+    for (int i = 0; i < 5; i++) {
```
Comment (Empact) [inline src/test/scheduler_tests.cpp]: nit: `++i` is preferred according to the developer-notes.

File: src/test/scheduler_tests.cpp [hunk 2]
```diff
@@ -112,4 +112,46 @@ BOOST_AUTO_TEST_CASE(manythreads)
...
+            BOOST_CHECK_EQUAL(i, counter2++);
+        });
+    }
+
+    // finish up
+    scheduler.stop(true);
+    threads.join_all();
+
+    BOOST_CHECK_EQUAL(counter1, 100);
+    BOOST_CHECK_EQUAL(counter2, 100);
```
Comment (Empact) [inline src/test/scheduler_tests.cpp]: Might want to label this a sanity check, to make it more clear that the 141/145 checks are the meat of the test.

File: src/validationinterface.h [hunk 3]
```diff
@@ -53,6 +53,19 @@ void CallFunctionInValidationInterfaceQueue(std::function<void ()> func);
...
 void SyncWithValidationInterfaceQueue();
 
+/**
+ * Implement this to subscribe to validation events
+ *
+ * Events are totally ordered - that is - all CValidationInterface() subscribers
+ * will receive callbacks in the same order. Furthermore, each ValidationInterface
+ * subscriber may assume that callbacks effectively run in a single thread with
+ * single-threaded memory consistency. That is, for a given ValidationInterface()
+ * instantiation, each callback will complete before
```
Comment (Empact) [inline src/validationinterface.h]: nit: whitespace

File: src/validationinterface.h [hunk 4]
```diff
@@ -53,6 +53,19 @@ void CallFunctionInValidationInterfaceQueue(std::function<void ()> func);
  */
 void SyncWithValidationInterfaceQueue();
 
+/**
+ * Implement this to subscribe to validation events
+ *
+ * Events are totally ordered - that is - all CValidationInterface() subscribers
```
Comment (TheBlueMatt) [inline src/validationinterface.h]: The wording here seems to imply order between different subscribers.

**Review by jamesob — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/13247/commits/9b6848e80eee6d65e6bfaf07188567c9cf37b394

File: src/scheduler.h [hunk 5]
```diff
@@ -86,9 +86,9 @@ class CScheduler
 
 /**
  * Class used by CScheduler clients which may schedule multiple jobs
- * which are required to be run serially. Does not require such jobs
- * to be executed on the same thread, but no two jobs will be executed
- * at the same time.
+ * which are required to be run serially. Jobs may not be run on the
+ * same thread, but no two jobs will be executed
+ * at the same time and memory will be sequentially consistent.
```
[the remark this answers is not in the document]
Comment (skeees) [reply src/scheduler.h]: updated, thanks

**Comment by skeees** [comment]
Updated to address @TheBlueMatt (diff is a comment change only)

**Comment by skeees** [comment]
Test failure was unrelated and now resolved

**Comment by jamesob** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/13247/commits/cbeaa91dbb1bc3ee6c05f3ee55a71268b8db2035

**Pull request merged**